/**
 * @file currentweatherwidget.cpp
 * @brief å®æ—¶å¤©æ°”å±•ç¤ºç»„ä»¶å®ç°
 */

#include "currentweatherwidget.h"
#include "ui_currentweatherwidget.h"
#include "../config/configmanager.h"

CurrentWeatherWidget::CurrentWeatherWidget(QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::CurrentWeatherWidget)
{
    ui->setupUi(this);
    setupConnections();
}

CurrentWeatherWidget::~CurrentWeatherWidget()
{
    delete ui;
}

void CurrentWeatherWidget::setupConnections()
{
    connect(ui->refreshBtn, &QPushButton::clicked,
            this, &CurrentWeatherWidget::onRefreshClicked);
}

void CurrentWeatherWidget::setCity(const QString &cityId, const QString &cityName)
{
    m_currentCityId = cityId;
    m_currentCityName = cityName;
    ui->cityNameLabel->setText(cityName);
    clear();
}

void CurrentWeatherWidget::updateWeather(const CurrentWeather &weather)
{
    if (!weather.isValid()) {
        return;
    }
    
    ConfigManager &config = ConfigManager::instance();
    
    // åŸå¸‚åç§°
    if (!weather.cityName.isEmpty()) {
        ui->cityNameLabel->setText(weather.cityName);
    }
    
    // æ¸©åº¦
    QString tempStr = config.formatTemperature(weather.temperature);
    ui->temperatureLabel->setText(tempStr);
    
    // ä½“æ„Ÿæ¸©åº¦
    QString feelsLikeStr = config.formatTemperature(weather.feelsLike);
    ui->feelsLikeLabel->setText(tr("ä½“æ„Ÿæ¸©åº¦ %1").arg(feelsLikeStr));
    
    // å¤©æ°”å›¾æ ‡å’Œæè¿°
    ui->weatherIconLabel->setText(getWeatherEmoji(weather.weatherCode));
    ui->weatherDescLabel->setText(weather.weatherDesc);
    
    // ç©ºæ°”è´¨é‡
    ui->aqiValueLabel->setText(tr("AQI %1").arg(weather.aqi));
    ui->aqiLabel->setText(getAqiLevel(weather.aqi));
    ui->aqiLabel->setStyleSheet(QString(
        "font-size: 14px; font-weight: bold; padding: 4px 12px; "
        "border-radius: 4px; background-color: %1; color: white;"
    ).arg(getAqiColor(weather.aqi)));
    
    // æ¹¿åº¦
    ui->humidityLabel->setText(QString("%1%").arg(weather.humidity));
    
    // é£é€Ÿå’Œé£å‘
    QString windStr = config.formatWindSpeed(weather.windSpeed);
    ui->windLabel->setText(windStr);
    ui->windDirLabel->setText(weather.windDirection);
    
    // æ°”å‹
    QString pressureStr = config.formatPressure(weather.pressure);
    ui->pressureLabel->setText(pressureStr);
    
    // èƒ½è§åº¦
    ui->visibilityLabel->setText(QString("%1 km").arg(weather.visibility));
    
    // æ—¥å‡ºæ—¥è½
    ui->sunriseLabel->setText(weather.sunriseTime.isEmpty() ? "--:--" : weather.sunriseTime);
    ui->sunsetLabel->setText(weather.sunsetTime.isEmpty() ? "--:--" : weather.sunsetTime);
    
    // æ›´æ–°æ—¶é—´
    QString updateStr = weather.updateTime.isValid() 
        ? weather.updateTime.toString("yyyy-MM-dd HH:mm")
        : "--";
    ui->updateTimeLabel->setText(tr("æ›´æ–°æ—¶é—´: %1").arg(updateStr));
}

void CurrentWeatherWidget::clear()
{
    ui->temperatureLabel->setText("--Â°");
    ui->feelsLikeLabel->setText(tr("ä½“æ„Ÿæ¸©åº¦ --Â°"));
    ui->weatherIconLabel->setText("â˜€");
    ui->weatherDescLabel->setText("--");
    ui->aqiLabel->setText("--");
    ui->aqiValueLabel->setText("AQI --");
    ui->humidityLabel->setText("--%");
    ui->windLabel->setText("-- km/h");
    ui->windDirLabel->setText("--");
    ui->pressureLabel->setText("-- hPa");
    ui->visibilityLabel->setText("-- km");
    ui->sunriseLabel->setText("--:--");
    ui->sunsetLabel->setText("--:--");
    ui->updateTimeLabel->setText(tr("æ›´æ–°æ—¶é—´: --"));
}

void CurrentWeatherWidget::onRefreshClicked()
{
    if (!m_currentCityId.isEmpty()) {
        emit refreshRequested(m_currentCityId);
    }
}

QString CurrentWeatherWidget::getWeatherEmoji(const QString &code)
{
    // å’Œé£å¤©æ°”å›¾æ ‡ä»£ç æ˜ å°„åˆ° emoji
    static QMap<QString, QString> emojiMap = {
        {"100", "â˜€ï¸"},   // æ™´
        {"101", "â›…"},   // å¤šäº‘
        {"102", "ğŸŒ¤ï¸"},  // å°‘äº‘
        {"103", "â›…"},   // æ™´é—´å¤šäº‘
        {"104", "â˜ï¸"},   // é˜´
        {"150", "ğŸŒ™"},   // æ™´(å¤œ)
        {"151", "ğŸŒ™"},   // å¤šäº‘(å¤œ)
        {"300", "ğŸŒ§ï¸"},  // é˜µé›¨
        {"301", "ğŸŒ§ï¸"},  // å¼ºé˜µé›¨
        {"302", "â›ˆï¸"},  // é›·é˜µé›¨
        {"303", "â›ˆï¸"},  // å¼ºé›·é˜µé›¨
        {"304", "â›ˆï¸"},  // é›·é˜µé›¨ä¼´æœ‰å†°é›¹
        {"305", "ğŸŒ§ï¸"},  // å°é›¨
        {"306", "ğŸŒ§ï¸"},  // ä¸­é›¨
        {"307", "ğŸŒ§ï¸"},  // å¤§é›¨
        {"308", "ğŸŒ§ï¸"},  // æç«¯é™é›¨
        {"309", "ğŸŒ§ï¸"},  // æ¯›æ¯›é›¨
        {"310", "ğŸŒ§ï¸"},  // æš´é›¨
        {"311", "ğŸŒ§ï¸"},  // å¤§æš´é›¨
        {"312", "ğŸŒ§ï¸"},  // ç‰¹å¤§æš´é›¨
        {"313", "ğŸŒ§ï¸"},  // å†»é›¨
        {"314", "ğŸŒ§ï¸"},  // å°åˆ°ä¸­é›¨
        {"315", "ğŸŒ§ï¸"},  // ä¸­åˆ°å¤§é›¨
        {"316", "ğŸŒ§ï¸"},  // å¤§åˆ°æš´é›¨
        {"400", "ğŸŒ¨ï¸"},  // å°é›ª
        {"401", "ğŸŒ¨ï¸"},  // ä¸­é›ª
        {"402", "â„ï¸"},   // å¤§é›ª
        {"403", "â„ï¸"},   // æš´é›ª
        {"404", "ğŸŒ¨ï¸"},  // é›¨å¤¹é›ª
        {"405", "ğŸŒ¨ï¸"},  // é›¨é›ªå¤©æ°”
        {"406", "ğŸŒ¨ï¸"},  // é˜µé›¨å¤¹é›ª
        {"407", "ğŸŒ¨ï¸"},  // é˜µé›ª
        {"500", "ğŸŒ«ï¸"},  // è–„é›¾
        {"501", "ğŸŒ«ï¸"},  // é›¾
        {"502", "ğŸŒ«ï¸"},  // éœ¾
        {"503", "ğŸŒ«ï¸"},  // æ‰¬æ²™
        {"504", "ğŸŒ«ï¸"},  // æµ®å°˜
        {"507", "ğŸŒ«ï¸"},  // æ²™å°˜æš´
        {"508", "ğŸŒ«ï¸"},  // å¼ºæ²™å°˜æš´
        {"900", "ğŸ”¥"},   // çƒ­
        {"901", "â„ï¸"},   // å†·
        {"999", "â“"}    // æœªçŸ¥
    };
    
    return emojiMap.value(code, "ğŸŒ¡ï¸");
}

QString CurrentWeatherWidget::getAqiColor(int aqi)
{
    if (aqi <= 50) return "#67C23A";       // ä¼˜ - ç»¿è‰²
    if (aqi <= 100) return "#E6A23C";      // è‰¯ - é»„è‰²
    if (aqi <= 150) return "#F56C6C";      // è½»åº¦æ±¡æŸ“ - æ©™è‰²
    if (aqi <= 200) return "#E6A23C";      // ä¸­åº¦æ±¡æŸ“ - çº¢è‰²
    if (aqi <= 300) return "#909399";      // é‡åº¦æ±¡æŸ“ - ç´«è‰²
    return "#303133";                       // ä¸¥é‡æ±¡æŸ“ - è¤çº¢è‰²
}

QString CurrentWeatherWidget::getAqiLevel(int aqi)
{
    if (aqi <= 50) return tr("ä¼˜");
    if (aqi <= 100) return tr("è‰¯");
    if (aqi <= 150) return tr("è½»åº¦æ±¡æŸ“");
    if (aqi <= 200) return tr("ä¸­åº¦æ±¡æŸ“");
    if (aqi <= 300) return tr("é‡åº¦æ±¡æŸ“");
    return tr("ä¸¥é‡æ±¡æŸ“");
}
